### PR TITLE
feat: Implements retrying logic for downloading models using --model-url flag

### PR DESCRIPTION
**Pull request by farbodbj**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

This PR implements the improvement requested at #6288, the approach is simple and uses a while loop for retrying with exponential backoff.
I found two calls to `curl_easy_perform` which is the [function responsible for transferring the data](https://curl.se/libcurl/c/curl_easy_perform.html), and on the call sites, I added the while loop for retrying on the condition of failing.
For now, the retry count is fixed as a local variable inside the function, but passing the retry count as a CLI argument can further improve this feature.
I compiled the project after the changes with `make LLAMA_CURL=1` and tested it both with and without internet connection and it worked okay.
It's my first time working with `curl` API in cpp so any suggestion for improvement is appreciated.

### CODE REVIEW

File: common/common.cpp [hunk 1]
```diff
@@ -2361,10 +2361,22 @@ static bool llama_download_file(const std::string & url, const std::string & pat
         curl_easy_setopt(curl.get(), CURLOPT_HEADERFUNCTION, static_cast<CURLOPT_HEADERFUNCTION_PTR>(header_callback));
         curl_easy_setopt(curl.get(), CURLOPT_HEADERDATA, &headers);
 
-        CURLcode res = curl_easy_perform(curl.get());
-        if (res != CURLE_OK) {
-            fprintf(stderr, "%s: curl_easy_perform() failed: %s\n", __func__, curl_easy_strerror(res));
-            return false;
+        int head_request_attempts = 3;
+        int remaining_request_attempts = 3;
+        int retry_delay = 2;
```
Comment (ngxson) [inline common/common.cpp:2366]: Value of `head_request_attempts` and `retry_delay` are fixed, you can extract it to a `#define`, for example:

```c
#define CURL_MAX_RETRY 3
#define CURL_RETRY_DELAY_SECONDS 2
```


```suggestion
        int remaining_request_attempts = CURL_MAX_RETRY;
```

File: common/common.cpp [hunk 2]
```diff
@@ -2361,10 +2361,22 @@ static bool llama_download_file(const std::string & url, const std::string & pat
...
+        while (remaining_request_attempts > 0){
+            CURLcode res = curl_easy_perform(curl.get());
+            if (res != CURLE_OK) {
+                int exponential_backoff_delay = std::pow(retry_delay, (head_request_attempts - remaining_request_attempts)) * 1000;
```
Comment (ngxson) [inline common/common.cpp:2370]: Then use the defined:

```suggestion
                int exponential_backoff_delay = std::pow(CURL_RETRY_DELAY_SECONDS, (CURL_MAX_RETRY - remaining_request_attempts)) * 1000;
```

**Review by ngxson — CHANGES_REQUESTED** [review]

**Comment by farbodbj** [comment]
@ngxson 
I applied your comments. wasn't sure where I had to put my `#define`s so I put them right above the function using them.

**Comment by farbodbj** [comment]
I would also appreciate if someone can help me to fix the `editorconfig-checker` CI job, I couldn't figure out how to run the job locally to fix formatting and white space problems

File: common/common.cpp [hunk 3]
```diff
@@ -2361,10 +2365,20 @@ static bool llama_download_file(const std::string & url, const std::string & pat
         curl_easy_setopt(curl.get(), CURLOPT_HEADERFUNCTION, static_cast<CURLOPT_HEADERFUNCTION_PTR>(header_callback));
         curl_easy_setopt(curl.get(), CURLOPT_HEADERDATA, &headers);
 
-        CURLcode res = curl_easy_perform(curl.get());
-        if (res != CURLE_OK) {
-            fprintf(stderr, "%s: curl_easy_perform() failed: %s\n", __func__, curl_easy_strerror(res));
-            return false;
+        int remaining_request_attempts = CURL_MAX_RETRY;
+        while (remaining_request_attempts > 0){
```
Comment (slaren) [inline common/common.cpp:2369]:
```suggestion
        while (remaining_request_attempts > 0) {
```

File: common/common.cpp [hunk 4]
```diff
@@ -2438,12 +2452,22 @@ static bool llama_download_file(const std::string & url, const std::string & pat
...
 
         // start the download
-        fprintf(stderr, "%s: downloading from %s to %s (server_etag:%s, server_last_modified:%s)...\n", __func__,
-                llama_download_hide_password_in_url(url).c_str(), path.c_str(), headers.etag.c_str(), headers.last_modified.c_str());
-        auto res = curl_easy_perform(curl.get());
-        if (res != CURLE_OK) {
-            fprintf(stderr, "%s: curl_easy_perform() failed: %s\n", __func__, curl_easy_strerror(res));
-            return false;
+        int remaining_attempts = CURL_MAX_RETRY;
+        while (remaining_attempts > 0){
```
Comment (slaren) [inline common/common.cpp:2456]:
```suggestion
        while (remaining_attempts > 0) {
```

**Review by slaren — APPROVED** [review]
I would suggest extracting the retry logic to a function rather than duplicating the code.

**Review by slaren — APPROVED** [review]

**Comment by slaren** [comment]
@ngxson Please update your review.

**Review by ngxson — APPROVED** [review]
Thanks. I'll merge when CI passes

**Pull request merged**